File: apps/backend/app/api/decisions.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, BackgroundTasks, Depends, Query, Response, status

from app.core.auth import CurrentUser
from app.core.config import Settings, get_settings
from app.models.decision import (
    ConversationTurn,
    DecisionCreate,
    DecisionCollectionUpdate,
    DecisionDetail,
    DecisionPage,
    DecisionMessageCreate,
    DecisionMessage,
    DecisionOption,
    DecisionOptionCreate,
    DecisionOptionUpdate,
    DecisionOptionStatusUpdate,
    DecisionStateItemReview,
    ContradictionResolution,
    DecisionSummary,
    DecisionTitleUpdate,
)
from app.services.decisions import DecisionStore


router = APIRouter(prefix="/decisions", tags=["decisions"])


@router.post("", response_model=DecisionSummary, status_code=status.HTTP_201_CREATED)
async def create_decision(
    values: DecisionCreate,
    background_tasks: BackgroundTasks,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionSummary:
    store = DecisionStore(settings, user)
    decision, user_message = await store.create_pending(values)
    background_tasks.add_task(store.process_initial_turn, decision.id, user_message)
    return decision


@router.get("", response_model=DecisionPage)
async def list_decisions(
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
    q: str | None = Query(default=None, max_length=200),
    collection_id: UUID | None = None,
    uncategorized: bool = False,
    trash: bool = False,
    cursor: str | None = None,
    limit: int = Query(default=20, ge=1, le=50),
) -> DecisionPage:
    return await DecisionStore(settings, user).list(
        search=q,
        collection_id=collection_id,
        uncategorized=uncategorized,
        trash=trash,
        cursor=cursor,
        limit=limit,
    )


@router.get("/{decision_id}", response_model=DecisionDetail)
async def get_decision(
    decision_id: UUID,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionDetail:
    return await DecisionStore(settings, user).get(decision_id)


@router.patch("/{decision_id}/title", response_model=DecisionSummary)
async def rename_decision(
    decision_id: UUID,
    values: DecisionTitleUpdate,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionSummary:
    return await DecisionStore(settings, user).rename(decision_id, values)


@router.post("/{decision_id}/messages", response_model=ConversationTurn, status_code=status.HTTP_201_CREATED)
async def add_decision_message(
    decision_id: UUID,
    values: DecisionMessageCreate,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> ConversationTurn:
    return await DecisionStore(settings, user).add_message(decision_id, values)


@router.post("/{decision_id}/options", response_model=DecisionOption, status_code=status.HTTP_201_CREATED)
async def create_decision_option(
    decision_id: UUID,
    values: DecisionOptionCreate,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionOption:
    return await DecisionStore(settings, user).create_option(decision_id, values)


@router.patch("/{decision_id}/options/{option_id}", response_model=DecisionOption)
async def update_decision_option(
    decision_id: UUID,
    option_id: UUID,
    values: DecisionOptionUpdate,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionOption:
    return await DecisionStore(settings, user).update_option(decision_id, option_id, values)


@router.delete("/{decision_id}/options/{option_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_decision_option(
    decision_id: UUID,
    option_id: UUID,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> Response:
    await DecisionStore(settings, user).delete_option(decision_id, option_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.patch("/{decision_id}/options/{option_id}/status", response_model=DecisionOption)
async def review_decision_option(
    decision_id: UUID,
    option_id: UUID,
    values: DecisionOptionStatusUpdate,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionOption:
    return await DecisionStore(settings, user).review_option(decision_id, option_id, values)


@router.patch("/{decision_id}/brief/items", response_model=DecisionDetail)
async def review_decision_state_item(
    decision_id: UUID,
    values: DecisionStateItemReview,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionDetail:
    return await DecisionStore(settings, user).review_state_item(decision_id, values)


@router.patch("/{decision_id}/brief/contradictions", response_model=DecisionDetail)
async def resolve_decision_contradiction(
    decision_id: UUID,
    values: ContradictionResolution,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionDetail:
    return await DecisionStore(settings, user).resolve_contradiction(decision_id, values)


@router.post("/{decision_id}/workflow/retry", response_model=DecisionMessage)
async def retry_decision_workflow(
    decision_id: UUID,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionMessage:
    return await DecisionStore(settings, user).retry_workflow(decision_id)


@router.put("/{decision_id}/collection", status_code=status.HTTP_204_NO_CONTENT)
async def move_decision(
    decision_id: UUID,
    values: DecisionCollectionUpdate,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> Response:
    await DecisionStore(settings, user).set_collection(decision_id, values)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{decision_id}/trash", response_model=DecisionSummary)
async def trash_decision(
    decision_id: UUID,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionSummary:
    return await DecisionStore(settings, user).trash(decision_id)


@router.post("/{decision_id}/restore", response_model=DecisionSummary)
async def restore_decision(
    decision_id: UUID,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> DecisionSummary:
    return await DecisionStore(settings, user).restore(decision_id)


@router.delete("/{decision_id}", status_code=status.HTTP_204_NO_CONTENT)
async def permanently_delete_decision(
    decision_id: UUID,
    user: CurrentUser,
    settings: Annotated[Settings, Depends(get_settings)],
) -> Response:
    await DecisionStore(settings, user).permanently_delete(decision_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: apps/frontend/src/lib/api/decisions.ts
import { createClient } from "@/lib/supabase/server";
import { redirect } from "next/navigation";

export type DecisionSummary = {
  id: string;
  title: string;
  prompt: string;
  status: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  collection_id: string | null;
  collection_name: string | null;
};

export type DecisionPage = { items: DecisionSummary[]; next_cursor: string | null };

export type Collection = {
  id: string;
  user_id: string;
  name: string;
  created_at: string;
  updated_at: string;
};

export type DecisionOption = {
  id: string;
  title: string;
  description: string | null;
  position: number;
  evaluation: Record<string, unknown>;
  source: "user_provided" | "ai_extracted" | "ai_generated";
  status: "candidate" | "confirmed" | "rejected";
  metadata: Record<string, unknown>;
};

export type DecisionMessage = {
  id: string;
  role: "user" | "assistant" | "system";
  content: string;
  created_at: string;
  structured_data: Record<string, unknown>;
};

export type Evaluation = {
  id: string;
  summary: string;
  confidence: number | null;
  risk_level: string | null;
  created_at: string;
};

export type DecisionDetail = DecisionSummary & {
  decision_brief: Record<string, unknown>;
  recommendation: Record<string, unknown> | null;
  options: DecisionOption[];
  messages: DecisionMessage[];
  evaluations: Evaluation[];
};

function apiUrl() {
  return process.env.API_URL ?? process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";
}

export async function authenticatedFetch<T>(path: string, init?: RequestInit, next = "/history"): Promise<T> {
  const supabase = createClient();
  const { data: { user } } = await supabase.auth.getUser();
  if (!user) redirect(`/auth?next=${encodeURIComponent(next)}`);

  const { data: { session } } = await supabase.auth.getSession();
  if (!session?.access_token) redirect("/auth");

  const response = await fetch(`${apiUrl()}${path}`, {
    ...init,
    cache: "no-store",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${session.access_token}`,
      ...init?.headers,
    },
  });

  if (response.status === 401) redirect("/auth");
  if (!response.ok) {
    const body = await response.json().catch(() => null);
    throw new Error(body?.detail ?? "The decision service request failed.");
  }
  if (response.status === 204) return undefined as T;
  return response.json() as Promise<T>;
}

export function listDecisions(filters: {
  q?: string;
  collectionId?: string;
  uncategorized?: boolean;
  trash?: boolean;
  cursor?: string;
} = {}) {
  const params = new URLSearchParams();
  if (filters.q) params.set("q", filters.q);
  if (filters.collectionId) params.set("collection_id", filters.collectionId);
  if (filters.uncategorized) params.set("uncategorized", "true");
  if (filters.trash) params.set("trash", "true");
  if (filters.cursor) params.set("cursor", filters.cursor);
  return authenticatedFetch<DecisionPage>(`/decisions?${params.toString()}`);
}

export function getDecision(id: string) {
  return authenticatedFetch<DecisionDetail>(`/decisions/${encodeURIComponent(id)}`);
}

export function createDecision(values: { prompt: string }) {
  return authenticatedFetch<DecisionSummary>("/decisions", {
    method: "POST",
    body: JSON.stringify(values),
  }, "/decision/conversation");
}

export function addDecisionMessage(id: string, content: string) {
  return authenticatedFetch<{ user_message: DecisionMessage; assistant_message: DecisionMessage }>(
    `/decisions/${encodeURIComponent(id)}/messages`,
    { method: "POST", body: JSON.stringify({ content }) },
    `/decision/${id}`,
  );
}

export function renameDecision(id: string, title: string) {
  return authenticatedFetch<DecisionSummary>(`/decisions/${encodeURIComponent(id)}/title`, {
    method: "PATCH",
    body: JSON.stringify({ title }),
  });
}

export function createDecisionOption(id: string, values: { title: string; description?: string }) {
  return authenticatedFetch<DecisionOption>(`/decisions/${encodeURIComponent(id)}/options`, {
    method: "POST",
    body: JSON.stringify(values),
  });
}

export function updateDecisionOption(
  decisionId: string,
  optionId: string,
  values: { title: string; description?: string },
) {
  return authenticatedFetch<DecisionOption>(
    `/decisions/${encodeURIComponent(decisionId)}/options/${encodeURIComponent(optionId)}`,
    { method: "PATCH", body: JSON.stringify(values) },
  );
}

export function deleteDecisionOption(decisionId: string, optionId: string) {
  return authenticatedFetch<void>(
    `/decisions/${encodeURIComponent(decisionId)}/options/${encodeURIComponent(optionId)}`,
    { method: "DELETE" },
  );
}

export function reviewDecisionOption(decisionId: string, optionId: string, status: "confirmed" | "rejected") {
  return authenticatedFetch<DecisionOption>(
    `/decisions/${encodeURIComponent(decisionId)}/options/${encodeURIComponent(optionId)}/status`,
    { method: "PATCH", body: JSON.stringify({ status }) },
  );
}

export function reviewDecisionBriefItem(
  decisionId: string,
  values: { collection: string; item_id: string; status: "confirmed" | "rejected"; replacement?: string },
) {
  return authenticatedFetch<DecisionDetail>(`/decisions/${encodeURIComponent(decisionId)}/brief/items`, {
    method: "PATCH",
    body: JSON.stringify(values),
  });
}

export function resolveDecisionContradiction(
  decisionId: string,
  values: { contradiction_id: string; resolution: "previous" | "new" | "custom"; custom_value?: string },
) {
  return authenticatedFetch<DecisionDetail>(`/decisions/${encodeURIComponent(decisionId)}/brief/contradictions`, {
    method: "PATCH",
    body: JSON.stringify(values),
  });
}

export function retryDecisionWorkflow(decisionId: string) {
  return authenticatedFetch<DecisionMessage>(`/decisions/${encodeURIComponent(decisionId)}/workflow/retry`, {
    method: "POST",
  });
}

export function listCollections() {
  return authenticatedFetch<Collection[]>("/collections");
}

export function createCollection(name: string) {
  return authenticatedFetch<Collection>("/collections", {
    method: "POST",
    body: JSON.stringify({ name }),
  });
}

export function renameCollection(id: string, name: string) {
  return authenticatedFetch<Collection>(`/collections/${encodeURIComponent(id)}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
  });
}

export function deleteCollection(id: string) {
  return authenticatedFetch<void>(`/collections/${encodeURIComponent(id)}`, { method: "DELETE" });
}

export function moveDecision(id: string, collectionId: string | null) {
  return authenticatedFetch<void>(`/decisions/${encodeURIComponent(id)}/collection`, {
    method: "PUT",
    body: JSON.stringify({ collection_id: collectionId }),
  });
}

export function trashDecision(id: string) {
  return authenticatedFetch<DecisionSummary>(`/decisions/${encodeURIComponent(id)}/trash`, { method: "POST" });
}

export function restoreDecision(id: string) {
  return authenticatedFetch<DecisionSummary>(`/decisions/${encodeURIComponent(id)}/restore`, { method: "POST" });
}

export function permanentlyDeleteDecision(id: string) {
  return authenticatedFetch<void>(`/decisions/${encodeURIComponent(id)}`, { method: "DELETE" });
}
